url/s3url.go

```diff
@@ -24,38 +40,112 @@ func (s S3Url) Format() string {
 	if s.Key == "" {
 		return s.Bucket
 	}
-	return s.Bucket + "/" + s.Key
+	return s.Bucket + s3Separator + s.Key
+}
+
+// setPrefixAndFilter sets prefix and filter parts on s3 key
+//
+// filter is the part that comes after the wildcard string
+// prefix is the part that comes before the wildcard string
+func (s *S3Url) setPrefixAndFilter() {
+	loc := strings.IndexAny(s.Key, s3WildCharacters)
+	wildOperation := loc > -1
+	if !wildOperation {
+		s.Delimiter = s3Separator
+		s.Prefix = s.Key
+	} else {
+		s.Prefix = s.Key[:loc]
+		s.filter = s.Key[loc:]
+	}
+}
+
+// setFilterRegex converts wildcard strings to regex format
+// and precompiles it for later usage. It is default to
+// ".*" to match every key on S3
+func (s *S3Url) setFilterRegex() error {
+	filterRegex := defaultRegex
+	if s.filter != "" {
+		filterRegex = regexp.QuoteMeta(s.filter)
+		filterRegex = strings.Replace(filterRegex, "\\?", ".", -1)
+		filterRegex = strings.Replace(filterRegex, "\\*", ".*?", -1)
+	}
+	filterRegex = s.Prefix + filterRegex
+	r, err := regexp.Compile("^" + filterRegex + "$")
+	if err != nil {
+		return err
+	}
+	s.filterRegex = r
+	return nil
 }
 
 // Clone creates a new s3url with the values from the receiver
 func (s S3Url) Clone() S3Url {
-	return S3Url{s.Bucket, s.Key}
+	return S3Url{
+		Bucket:      s.Bucket,
+		Delimiter:   s.Delimiter,
+		Key:         s.Key,
+		Prefix:      s.Prefix,
+		filter:      s.filter,
+		filterRegex: s.filterRegex,
+	}
+}
+
+// Match check if given key matches with regex and
+// returns parsed key
+func (s S3Url) Match(key string) (string, bool) {
+	if !s.filterRegex.MatchString(key) {
+		return "", false
+	}
+
+	var index int
+	// if it is a batch operation, get key before the filter part
+	// starting from last "/", otherwise get last directory or file
+	isBatch := s.filter != ""
+	if isBatch {
+		index = strings.LastIndex(s.Prefix, s3Separator)
+	} else {
+		// trim last char if it is "/"
+		trim := strings.TrimSuffix(key, s3Separator)
+		index = strings.LastIndex(trim, s3Separator)
+	}
+	return trimKey(key, index), true
+}
+
+// trimKey trims keys by given index and clears
+// "/" prefix if it exists
+func trimKey(key string, index int) string {
+	if index < 0 {
+		index = 0
+	}
+	trimmedKey := key[index:]
+	trimmedKey = strings.TrimPrefix(trimmedKey, s3Separator)
+	return trimmedKey
 }
 
 // HasWild checks if a string contains any S3 wildcard chars
 func HasWild(s string) bool {
-	return strings.ContainsAny(s, S3WildCharacters)
+	return strings.ContainsAny(s, s3WildCharacters)
 }
 
 // ParseS3Url parses a string into an S3Url
 func ParseS3Url(object string) (*S3Url, error) {
```

Comment:
now that we're here, we should call this func `New`.

Reply:
done